File: terminator-application/src/test/java/io/quartic/terminator/CatalogueWatcherShould.java
package io.quartic.terminator;

import com.fasterxml.jackson.databind.JavaType;
import com.google.common.collect.ImmutableMap;
import io.quartic.catalogue.api.*;
import io.quartic.common.client.WebsocketListener;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.Map;
import java.util.Optional;

import static java.util.Collections.emptyMap;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.empty;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static rx.Observable.just;

public class CatalogueWatcherShould {
    private final WebsocketListener<Map<DatasetId, DatasetConfig>> listener = mock(WebsocketListener.class);
    private final WebsocketListener.Factory listenerFactory = mock(WebsocketListener.Factory.class);
    private final CatalogueWatcher proxy = CatalogueWatcher.of(listenerFactory);

    @Before
    public void before() throws Exception {
        when(listenerFactory.create(any(JavaType.class))).thenReturn((WebsocketListener)listener);
    }

    @After
    public void after() throws Exception {
        proxy.close();
    }

    @Test
    public void expose_returned_ids() throws Exception {
        final TerminationId terminationId = TerminationIdImpl.of("456");
        final ImmutableMap<DatasetId, DatasetConfig> datasets = datasetsWithLocator(TerminatorDatasetLocatorImpl.of(terminationId));
        when(listener.observable()).thenReturn(just(datasets));

        proxy.start();

        assertThat(proxy.terminationIds(), contains(terminationId));
    }

    @Test
    public void ignore_incorrect_types() throws Exception {
        final ImmutableMap<DatasetId, DatasetConfig> datasets = datasetsWithLocator(PostgresDatasetLocatorImpl.of("a", "b", "c", "d"));
        when(listener.observable()).thenReturn(just(datasets));

        proxy.start();

        assertThat(proxy.terminationIds(), empty());
    }

    private ImmutableMap<DatasetId, DatasetConfig> datasetsWithLocator(DatasetLocator locator) {
        return ImmutableMap.of(
                DatasetIdImpl.of("123"),
                DatasetConfigImpl.of(
                        DatasetMetadataImpl.of("foo", "bar", "baz", Optional.empty()),
                        locator,
                        emptyMap()
                ));
    }
}

File: terminator-application/src/test/kotlin/io/quartic/terminator/CatalogueWatcherShould.kt
package io.quartic.terminator

import com.fasterxml.jackson.databind.JavaType
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import io.quartic.catalogue.api.*
import io.quartic.common.websocket.WebsocketListener
import org.hamcrest.Matchers.contains
import org.hamcrest.Matchers.empty
import org.junit.After
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import rx.Observable.just
import java.time.Instant
import java.util.*
import java.util.Collections.emptyMap

class CatalogueWatcherShould {
    private val listener = mock<WebsocketListener<Map<DatasetId, DatasetConfig>>>()
    private val listenerFactory = mock<WebsocketListener.Factory>()
    private val proxy = CatalogueWatcher(listenerFactory)

    @Before
    fun before() {
        whenever(listenerFactory.create<Map<DatasetId, DatasetConfig>>(any<JavaType>())).thenReturn(listener)
    }

    @After
    fun after() {
        proxy.close()
    }

    @Test
    fun expose_returned_ids() {
        @Suppress("USELESS_CAST")
        val terminationId = TerminationIdImpl.of("456") as TerminationId
        val datasets = datasetsWithLocator(TerminatorDatasetLocatorImpl.of(terminationId))
        whenever(listener.observable).thenReturn(just(datasets))

        proxy.start()

        assertThat(proxy.terminationIds, contains(terminationId))
    }

    @Test
    fun ignore_incorrect_types() {
        val datasets = datasetsWithLocator(PostgresDatasetLocatorImpl.of("a", "b", "c", "d"))
        whenever(listener.observable).thenReturn(just(datasets))

        proxy.start()

        assertThat(proxy.terminationIds, empty())
    }

    private fun datasetsWithLocator(locator: DatasetLocator): Map<DatasetId, DatasetConfig> = hashMapOf(DatasetIdImpl.of("123") to DatasetConfigImpl.of(
            DatasetMetadataImpl.of(
                    "foo",
                    "bar",
                    "baz",
                    Optional.of(Instant.now()),
                    Optional.empty<Icon>()
            ),
            locator,
            emptyMap<String, Any>()
    ))
}
